### ISSUE TITLE
Model losses all predictive power in the first steps during fine tuning

### ISSUE THREAD

**Issue by jumpyknight** [google/deepvariant#185]
I'm trying to fine tune the original DeepVariant model with some extra data.
However, during the fine tuning process, the model suddenly losses all its predictive power in the first 10000 or 20000 steps. The call-variants output of these models are like all sites have homo-alt variants with a same qual value, 6.8 for example.
The sudden change in the model happens in the first step of fine tuning, as the saved model.ckpt-0 in the begining already gives the above output.
I find this result quite confusing, as the loss should increase dramatically. Is that a normal output of the fine tuning process?

**Comment by pichuan** [comment]
Hi @jumpyknight ,

We have also noticed that when warmstarting from a checkpoint, there seems to be dip in accuracy at the beginning. You can see an example like this in this plot in our training case study from r0.7:
https://raw.githubusercontent.com/google/deepvariant/r0.7/docs/images/TensorBoardAccuracy.png

And specifically, for a while it was puzzling to us why model.ckpt-0 was much less accurate.

After looking much closely into TensorFlow and the behavior or warmstarting, here is what we found:

It turns out that there's some subtlety to what exactly gets loaded in warm starting. It's documented here:
https://www.tensorflow.org/api_docs/python/tf/train/warm_start
Specifically, this:
```
vars_to_warm_start: [Optional] One of the following:
 * A regular expression (string) that captures which variables to warm-start (see tf.get_collection). This expression will only consider variables in the TRAINABLE_VARIABLES collection -- if you need to warm-start non_TRAINABLE vars (such as optimizer accumulators or batch norm statistics), please use the below option.
```
Because in our code, we use a regular expression like this:
https://github.com/google/deepvariant/blob/aff131aac3bd0cb63ee8314e32bcbf5590987fb8/deepvariant/modeling.py#L361

This means that only the variables in the TRAINABLE_VARIABLES get loaded when we warm start.
This explains why warm starting from a checkpoint does not produce the same (or even similar) results as the inference/prediction mode. For example, Batch norms were not being warmstarted.

If you do want to load **everything**,  one trick we've tried is to change to code into a list:
```
          vars_to_warm_start=['|'.join(vars_to_include)])
```
which then tricks the code into the model of loading everything (not just TRAINABLE_VARIABLES).

But, it doesn't seem very desirable because it also loads things like global_step.

The best way is probably to include just BatchNorm and the Trainable variables we want, but it'll require more refactoring here.

If try with the small code change above and observe the log -- you'll see that a lot more vars are being used in warm-starting now. (But note that you'll likely want to exclude some variables such as global_step, because those will affect your learning rate)

Note that even when warmstarting from an existing checkpoint, we don't currently recommend "fine-tuning" with very few steps. Even with our small TPU training case study, we showed an example to train up to 50k steps before we pick a model.
So if you want to use our code to fine tune on your data, please proceed with caution and make sure you carefully benchmark the accuracy and robustness of the resulting model. 

Thanks for sharing your findings here. If you have more questions or anything you want to discuss, please feel free to bring up here.

(And, adding my teammate @emschorsch who has been looking into warmstarting. )

**Comment by jumpyknight** [comment]
Thanks for the reply. I think it solves my problem.
I also agree that loading all variables is not the best, but I'd like to try the suggested code and check if the model would be the same first.
But still I think some vars like the EMA ones should be loaded in the warm-up stage, and I'll try to figure out what vars are needed.
I'll reply here if I make any further progress.

**Comment by pichuan** [comment]
Great! Looking forward to hearing what you find. I'll close this issue. Feel free to update here or open another issue if you encounter other problems.
